### PR TITLE
Migrate `environments` api from rails to spark

### PR DESCRIPTION
**Pull request by vishaldevgire**
Issue: #5698 

- Fixes performance degradation experienced since 18.10.0 release
Results (Tested with 500+ pipelines to emulate the issue):
   - For 18.10.0: 
`time curl 'https://localhost:8082/go/api/admin/environments'   0.03s user 0.02s system 2% cpu 2.484 total`
   - Now: `time curl 'https://localhost:8154/go/api/admin/environments'   0.03s user 0.02s system 27% cpu 0.191 total`

- Added a toggle to switch back to `rails` API, default API is `spark` from now:
   `curl http://localhost:8153/go/api/admin/feature_toggles/use_old_environments_api -d toggle_value=on -H 'Confirm:true' -u admin:badger`

### CODE REVIEW

File: api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/representers/PatchEnvironmentRequestRepresenter.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    public static PatchEnvironmentRequest fromJSON(JsonReader jsonReader) {
+        List<String> pipelineToAdd = extractListFromJson(jsonReader, "pipelines", "add");
+        List<String> pipelineToRemove = extractListFromJson(jsonReader, "pipelines", "remove");
+        List<String> agentsToAdd = extractListFromJson(jsonReader, "agents", "add");
+        List<String> agentsToRemove = extractListFromJson(jsonReader, "agents", "remove");
+        List<String> envVariablesToRemove = extractListFromJson(jsonReader, "environment_variables", "remove");
+
+        List<EnvironmentVariableConfig> environmentVariablesToAdd = new ArrayList<>();
+
+        if (jsonReader.hasJsonObject("environment_variables")) {
```
Comment (ketan) [inline api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/representers/PatchEnvironmentRequestRepresenter.java]: See `BulkUpdateRequestRepresenter` for a better way of writing this.

Comment (vishaldevgire) [reply api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/representers/PatchEnvironmentRequestRepresenter.java]: Incorporated the suggestion.

File: api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/EnvironmentsControllerV2.java [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+        throw haltBecauseEntityAlreadyExists(jsonWriter(environmentConfig), "environment", environmentConfig.name().toString());
+    }
+
+    private String calculateEtag(List<EnvironmentConfig> environmentConfigs) {
+        final String environmentConfigSegment = environmentConfigs
+                .stream()
+                .map(this::etagFor)
+                .collect(Collectors.joining(SEP_CHAR));
+
+        return DigestUtils.sha256Hex(StringUtils.joinWith(SEP_CHAR, currentUsername().getUsername(), environmentConfigSegment));
```
Comment (GaneshSPatil) [inline api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/EnvironmentsControllerV2.java]: @vishaldevgire @kritika-singh3 -- This ETag Computation has a problem.

Environments can be viewed only by super admins. And the view of the environments doesn't change based on the super administrator accessing it (unlike dashboard).
Adding the user information while computing ETag will result into different ETags based on the user who is accessing it.

IMO, the ETag computation logic for environments should be user agnostic. And should result in same ETag.

File: server/config/cruise-config.xml [hunk 3]
```diff
@@ -43,4 +43,7 @@
       </stage>
     </pipeline>
   </pipelines>
+  <agents>
```
Comment (GaneshSPatil) [inline server/config/cruise-config.xml]: Not required?

Comment (vishaldevgire) [reply server/config/cruise-config.xml]: added by mistake, removing it.

File: api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/EnvironmentsControllerV2.java (hunk 2 above)
Comment (vishaldevgire) [reply api/api-environments-v2/src/main/java/com/thoughtworks/go/apiv2/environments/EnvironmentsControllerV2.java]: Corrected.

**Pull request merged**